### PR TITLE
Make check of pixel array size more strict to conform to RDD 52 (new PR)

### PR DESCRIPTION
**Pull request by matmat**
New PR, because I messed up my branches. Sorry!

### CODE REVIEW

**Comment by remia** [comment]
Looks good to me.

Same comment as before, would it be safe to merge as is and update later when we have a dedicated way of handling IOP / SMPTE differences ? I guess I'm just checking if merging will not make a lot of DCP suddenly fail, or maybe it has been standard practice to use only these resolution for such a long time that it wouldn't matter ?

**Comment by matmat** [comment]
Maybe a compromise would be to emit a warning for 2048x1080/4096x2160 and 1920x1080/3840x2160, while we make array sizes in the "Other" category an error?

Here ar some statistics of pixel array sizes for the films at the last couple of years for [Göteborg Film Film Festival](https://goteborgfilmfestival.se/) where I manage the prints. The films are a mix of new (at most one year old) titles from big sale companies/studios and independently produced films.

|Year | 1998x1080 / 3996x2160 | 2048x858 / 4096x1716 | 2048x1080 / 4096x2160 | 1920x1080 / 3840x2160 | Other | Total |
|---|:---:|:---:|:---:|:---:|:---:|:---:|
|2013|145 (43%) |107 (32%)|35 (10%)|45 (13%)|4 (1%)|336|
|2014|191 (44%)|145 (33%)|34 (8%)|61 (14%)|3 (1%)|434|
|2015|204 (47%)|141 (32%)|23 (5%)|68 (16%)|1 (0%)|437|
|2016|247 (53%)|136 (29%)|17 (4%)|64 (14%)|3 (1%)|467|
|2017|266 (59%)|152 (34%)|16 (4%)|19 (4%)|0 (0%)|453|
|2018|293 (66%)|128 (29%)|12 (3%)|12 (3%)|0 (0%)|445|
|2019|260 (59%)|159 (36%)|19 (4%)|6 (1%)|0 (0%)|444|
|2020|237 (64%)|117 (32%)|11 (3%)|6 (2%)|0 (0%)|371|
|2021|49 (70%)|16 (23%)|4 (6%)|1 (1%)|0 (0%)|70|
|**Average**|**1892 (55%)**|**1101 (32%)**|**171 (5%)**|**282 (8%)**|**11 (0%)**|**3457**|

I know for a fact that som firmware versions of Doremi/Dolby IMS1000/IMS2000/IMS3000 have problems playing DCP:s in the "Other" category.

**Comment by remia** [comment]
Really interesting to see those statistics, thanks for sharing @matmat !

> Maybe a compromise would be to emit a warning for 2048x1080/4096x2160 and 1920x1080/3840x2160, while we make array sizes in the "Other" category an error?

That sounds like a good default behaviour to me.

**Comment by jamiegau** [comment]
from memory and being at the ISDCF meetings and these issues being discussed, Flat/Scope/Fill-Container are suppose to be the target frame sizes.  And anything non-standard should be placed in the best suitable container with black bars/pillars.  And is why DCNC supports the F-185 nomenclature.

I am not sure if this is formalised in a document.  Best to ask Jerry, the Chair, he would remember.

**Comment by matmat** [comment]
It is kind of a mess.

The pixel array size is specified in many places:

## [SMPTE RDD 52:2020](https://doi.org/10.5594/SMPTE.RDD52.2020)
> ### 7.1 Picture Essence Encoding
> For picture essence tracks, the pixel array size, frame rate, and 2D/3D shall be one of the formats listed in Table 1 below.
>
> Source images having an aspect ratio not listed in Table 1 below shall be encoded so that the image fills either the horizontal or vertical dimension of the desired SCOPE or FLAT Container (2K or 4K). To fill the pixel array in the opposite dimension, the image should be padded with an equal number of black pixels on each side, i.e., "letter-box" (top side, bottom side) or "pillar-box" (left side, right side).
>
> **Table 1 — Pixel Array Dimensions**<br>
>|Format|Horizontal Pixels|Vertical Pixels|Frame Rate|2D/3D|
>|:--:|:--:|:--:|:--:|:--:|
>|2K Scope (2.39:1)|2048|858|24/1, 25/1, or 48/1|2D/3D|
>|2K Flat (1.85:1)|1998|1080|24/1, 25/1, or 48/1|2D/3D|
>|4K Scope (2.39:1)|4096|1716|24/1|2D ONLY|
>|4K Flat (1.85:1)|3996|2160|24/1|2D ONLY|
>
> _NOTE: 4K 3D is only supported on a very limited number of systems in presentation environments such as museums, and special venues. Wide industry support does not yet exist, as it’s based on equipment capabilities. Use at your own risk._
>
> For purposes of this Application, 2K Full (1.90:1 — 2048x1080) and 4K Full (1.9:1 — 4096x2160) container size shall not be utilized.
>
> ## 7.1.1 Active Image Area
> For full active image area resolutions SCOPE or FLAT, image shall fill the frame edge to edge in both vertical and horizontal directions. 3D “floating windows” are acceptable but must be within the edge boundaries.
>
> All other aspect ratios (1.90, 1.40, 1.78, etc) and resolutions shall be sized and padded with black to fill the frame to the SCOPE or FLAT container resolutions in Table 1 above. The closest container to content’s frame size should be utilized to take advantage of as many pixels as possible to preserve quality. See Table 2 below for some examples of common padding of active image to FLAT and SCOPE containers.
> 
>**Table 2 - Common Active Image Sizes and Padding**
>|Active Image Aspect Ratio|Container Aspect Ratio|Active Image Size (2K/4K)|DCI Container Size (2K/4K)|
>|:--:|:--:|:--:|:--:|
>|SCOPE/2.39|FLAT/1.85|1998x836/3996x1672<sup>2</sup>|1998x1080/3996x2160|
>|FLAT/1.85|SCOPE/2.39|1588x858/3176x1716<sup>3</sup>|2048x858/4096x1716|
>|“Full Container’/1.90|FLAT/1.85|1998x1051/3996x2102|1998x1080/3996x2160|
>|HD/1.78|FLAT/1.85|1922x1080/3844x2160|1998x1080/3996x2160|
>|CinemaScope/2.35|SCOPE/2.39|2016x858/4032x1716|2048x858/4096x1716|
>|Traditional TV/1.33 (4:3)|FLAT/1.85|1436x1080/2872x2160|1998x1080/3996x2160|
>|Academy Ratio/1.375|FLAT/1.85|1485x1080/2970x2160|1998x1080/3996x2160|
>|Paramount Format/1.66|FLAT/1.85|1792x1080/3585x2160|1998x1080/3996x2160|
>
> 2 For SCOPE trailers playing in front of FLAT features.
> 3 For FLAT trailers playing in front of SCOPE features.

## [SMPTE ST 429-2:2020](https://doi.org/10.5594/SMPTE.ST429-2.2020)
> ### 8.2 Picture Essence Encoding
>Picture essence tracks shall be encoded as specified in [SMPTE ST 428-1](https://doi.org/10.5594/SMPTE.ST428-1.2019). The pixel array size and frame rate shall be one of the formats listed in Table 1. Monoscopic picture essence tracks shall have matching frame rate and edit rate. 
Stereoscopic picture essence tracks shall be limited to the 2K formats, and shall have a frame rate of 48/1 and an edit rate equal to half the frame rate (_r<sub>e</sub>  = r<sub>f</sub> / 2_). (See [SMPTE ST 429-10](https://doi.org/10.5594/SMPTE.ST429-10.2008) for an explanation).
>
> Source images having an aspect ratio not listed in Table 1 should be encoded so that the image fills either the horizontal or vertical dimension of the desired Full pixel array (2K or 4K). To fill the pixel array in the opposite dimension, the image should be padded with an equal number of black pixels on each side, i.e., "letter-box" (top side, bottom side) or "pillar-box" (left side, right side).
>
> **Table 1 – Pixel Array Dimensions**
>| Format | Horizontal Pixels | Vertical Pixels | Frames rate |
>|:---|:---:|:---:|:---:|
>|_2K Scope (2.39:1)_|2048|858|24/1, 25/1, 30/1, 48/1, 50/1 or 60/1|
>|_2K Flat (1.85:1)_|1998|1080|24/1, 25/1, 30/1, 48/1, 50/1 or 60/1|
>|_2K Full (1.90:1)_|2048|1080|24/1, 25/1, 30/1, 48/1, 50/1 or 60/1|
>|_4K Scope (2.39:1)_|4096|1716|24/1, 25/1 or 30/1|
>|_4K Flat (1.85:1)_|3996|2160|24/1, 25/1 or 30/1|
>|_4K Full (1.90:1)_|4096|2160|24/1, 25/1 or 30/1|

## [SMPTE ST 428-1:2006](https://doi.org/10.5594/SMPTE.ST428-1.2006)
(There is a [428-1:2019](https://doi.org/10.5594/SMPTE.ST428-1.2019) but unfortunately I don't have access to it):
> ### 3 Image structures
> (...) The maximum number of horizontal and vertical pixels shall be no greater than the limits of one of the operational levels as defined in Table 12. The number of pixels shall extend to the maximum in either one or both of the horizontal and vertical directions of the defined operational level.
>
> **Table 1 – DCDM operational levels**
>| Operational Level | Maximum Horizontal Pixels | Maximum Vertical Pixels | Frames per Second |
>|:---:|:---:|:---:|:---:|
>|1|4096|2160|24|
>|2|2048|1080|48|
>|3|2048|1080|24|
>
> (...)
> For operational level 1, the number of horizontal and vertical pixels shall be evenly divisible by four.
> For operational levels 2 and 3, the number of horizontal and vertical pixels shall be evenly divisible by two.


## [SMPTE ST 428-11:2013](https://doi.org/10.5594/SMPTE.ST428-11.2013)
> ### 5 Additional Frame Rate Values
> The primary D-Cinema frame rates and levels shall be as defined by [SMPTE ST 428-1](https://doi.org/10.5594/SMPTE.ST428-1.2019).
>
> The additional frame rate values for D-Cinema use shall be 25, 30, 50, 60, 96, 100 and 120 frames per second. All additional frame rates shall be integer values.
>
> ### 5.1. Definition of AFR levels
> The additional frame rates shall be defined in association with the picture size to form AFR levels as shown in Table 1.
> 
> **Table 1 – AFR Level Definitions**
> |AFR Level|Maximum Horizontal Pixels|Maximum Vertical Pixels|Frames per Second|
> |:--:|:--:|:--:|:--:|
> |AFR Level 1|4096|2160|25|
> |AFR Level 2|2048|1080|25|
> |AFR Level 3|4096|2160|30|
> |AFR Level 4|2048|1080|30|
> |AFR Level 5|2048|1080|50|
> |AFR Level 6|2048|1080|60|
> |AFR Level 7|2048|1080|96|
> |AFR Level 8|2048|1080|100|
> |AFR Level 9|2048|1080|120|
>
> These AFR fevels are in addition to the primary DCDM levels that are detined in [SMPTE ST 428-1](https://doi.org/10.5594/SMPTE.ST428-1.2019).

## [SMPTE ST 428-21:2011](https://doi.org/10.5594/SMPTE.ST428-21.2011)
> ## 5 Archive Frame Rate Values
> This standard supplements the provisions of  [SMPTE ST 428-1](https://doi.org/10.5594/SMPTE.ST428-1.2019) with archive frame rates and levels.
>
> The archive frame rate values for D-Cinema use shall be 16/1, 200/11, 20/1 and 240/11 frames per second.
>
> ## 5.1 Definition of Archive Levels
> The Archive Level shall have eight values defined as Archive Level 1 through Archive Level 8. The picture container size and frame rate for each Archive Level shall be as defined in Table 1:
>
> **Table 1 — Archive Level Definitions**
>|Archive Level|Maximum Horizontal Pixels|Maximum Vertical Pixels|Frames per Second|
>|:--:|:--:|:--:|:--:|
>|Archive Level 1|4096|2160|16/1|
>|Archive Level 2|2048|1080|16/1|
>|Archive Level 3|4096|2160|200/11|
>|Archive Level 4|2048|1080|200/11|
>|Archive Level 5|4096|2160|20/1|
>|Archive Level 6|2048|1080|20/1|
>|Archive Level 7|4096|2160|240/11|
>|Archive Level 8|2048|1080|240/11|
>
> These archive levels shall be in addition to the DCDM levels that are defined in [SMPTE ST 428-1](https://doi.org/10.5594/SMPTE.ST428-1.2019).


## [SMPTE ST 429-13:2009](https://doi.org/10.5594/SMPTE.ST429-13.2009)
> ### 7.2 Picture Essence Encoding
> Picture essence encoding shall be as specified in [SMPTE 428-1](https://doi.org/10.5594/SMPTE.ST428-1.2019).
>
> For each of the AFRs listed, the pixel array size and frame rate shall be one of the formats listed in Table 1. Monoscopic picture essence tracks shall have matching frame rate and edit rate. Stereoscopic picture essence tracks shall be limited to the 2K formats and shall have the frame rates of 50/1 or 60/1 with an edit rate equal to half the frame rate (_r<sub>e</sub>  = r<sub>f</sub> / 2_).
> 
> Source images having an aspect ratio not listed in Table 1 should be encoded so that the image fills either the horizontal or vertical dimension of the desired full pixel array (2K or 4K). To fill the pixel array in the opposite dimension, the image should be padded with an equal number of black pixels on each side, i.e., "letter-box" (top side, bottom side) or "pillar-box" (left side, right side).
>
> **Table 1 – Pixel Array Dimensions**
>| Format | Horizontal Pixels | Vertical Pixels | Frames rate |
>|:---|:---:|:---:|:---:|
>|2K Scope (2.39:1)|2048|858|25/1, 30/1, 50/1 or 60/1|
>|2K Flat (1.85:1)|1998|1080|25/1, 30/1, 50/1 or 60/1|
>|2K Full (1.90:1)|2048|1080|25/1, 30/1, 50/1 or 60/1|
>|4K Scope (2.39:1)|4096|1716|25/1 or 30/1|
>|4K Flat (1.85:1)|3996|2160|25/1 or 30/1|
>|4K Full (1.90:1)|4096|2160|25/1 or 30/1|

## [mpeg_ii_cpl_spec.doc (v2.4)](https://interop-docs.cinepedia.com/Document_Release_2.0/mpeg_ii_cpl_spec.pdf)
> ### 6.4.2 ScreenAspectRatio
>
> The ScreenAspectRatio parameter reflects the aspect ratio of the picture information contained in the underlying picture track file. As such, it is included in the composition playlist for convenience only and may be used, for instance, to select an appropriate screen mask. It is supplemented by an optional scope attribute. Unless the scope attribute is specified, the content of the field shall be one of the values described in Table 10.
> **Table 10. Screen Aspect Ratios.**
>
> | |
> |--|
> |1.33|
> |1.66|
> |1.77|
> |1.85|
> |2.00|
> |2.39|

## [mpeg_ii_track.doc (v2.7)](https://interop-docs.cinepedia.com/Document_Release_2.0/mpeg_ii_track.pdf)
> ### 9.1.1 Picture Parameters
> Raster resolution shall be 1920 by 1080.

**Comment by matmat** [comment]
The idea that that the number of pixels in a valid track file pixel array should:

> _Always extend to the maximum in either one or both of the horizontal and vertical directions of the full 2048x1080/4096x2160 container._ 

Is actually a good idea! But unfortunately only in theory. It does not work out so well in practice.

There are really two different aspects to take into account here:

**The technical**
I.e. Does it play at all? As I said above some mediablocks with certain fw versions refuse to play content with unusual pixel arrays, even if compliant with the above. I have had mediablocks refuse to play a picture track with a 1480x1080 pixel array for example.


**The "soft" asptects:**
Most auditoriums have their projector lens settings not adjusted for anything other than 1.85:1 and 2.39:1.

Playing back e.g. a DCP with 2.20:1 active image area stored within a 2048x930 pixel array will not screen correctly at most places (it will be excessivly cropped) except at the very few cinemas with skilled projection staff that able to set projector correctly. I believe that RDD 52 is a testament to the pragmatic realisations of this (sad) fact.

Also look at the flood of released titles in 2.00:1 aspect ratio recently:

These films are distributed packaged with an active image area of 1998x1000 within a 1998x1080 pixel array, _**not**_ in 2048x1024 which would be appropriate if the standards should have been followed.

Another prominent example is [DUNKIRK](https://www.imdb.com/title/tt5013056/) (2017), this title was distributed with an active image area of 3996x1818 (2.20:1) within a 3996x2160 (1.85:1) pixel array, _**not**_ within a 4096x1860 (2.20) pixel array.

In conclusion, I think that the RDD 52 actually offer sensible advise on this matter: I.e. to accept only Flat and Scope containers. It is a bit sad, but the harsh reality is that these are the only two formats that are going to consistently screened correctly everywhere.

**Comment by remia** [comment]
Would it be a good compromise to split into two checks (or now two sub errors), one being the current behaviour and the other the more strict RDD52 recommendation ? That way we could have better control of error levels using the check profile.

**Comment by jamiegau** [comment]
Just a note here..  I have been working on a new free to use tool for cinemas that is a LMS (Library Management System) for cinemas.  Used to ingest and pull in DCPs from different services and as part of it, test all incoming DCPs using clairmeta.  I.e. when a DCP arrives, clairmeta is applied against the DCP to ensure any issues are exposed as soon as possible giving the cinema more time to react to any data corruption or other issues that may arise.

I am on clairmeta 1.1.1 currently, but am keen to upgrade to the next major release (Due to some false positives) before I make the tool widely available.

Do you have a estimation on when that will be?
